src/QsCompiler/Transformations/LiftLambdas.cs

```
@@ -0,0 +1,257 @@
+                        ? new CallableInformation(ResolvedCharacteristics.Empty, InferredCallableInformation.NoInformation)
+                        : throw new ArgumentException("Lambda with non-callable type");
+
+                    // Returns are allowed only if we are not doing an adjoint specialization.
+                    var isReturnStatement =
+                       !lambdaBody.Expression.IsUnitValue // There will be no statements in this case.
+                       && (callableInfo.Characteristics.SupportedFunctors.IsNull
+                       || !callableInfo.Characteristics.SupportedFunctors.Item.Contains(QsFunctor.Adjoint));
+                    var bodyStatment = new QsStatement(
+                        isReturnStatement
+                            ? QsStatementKind.NewQsReturnStatement(lambdaBody)
+                            : QsStatementKind.NewQsExpressionStatement(lambdaBody),
+                        LocalDeclarations.Empty,
+                        QsNullable<QsLocation>.Null,
+                        QsComments.Empty);
+                    var generatedContent = new QsScope(
+                        lambdaBody.Expression.IsUnitValue
+                            ? ImmutableArray<QsStatement>.Empty // if it is just a single unit literal, there should be an empty body
+                            : new[] { bodyStatment }.ToImmutableArray(),
+                        new LocalDeclarations(this.SharedState.KnownVariables));
+
+                    var success = processedLambda.Kind.IsFunction
+                        ? ContentLifting.LiftContent.LiftFunctionBody(this.SharedState.CurrentCallable!, generatedContent, lambdaParams, isReturnStatement, out var call, out var callable)
+                        : ContentLifting.LiftContent.LiftOperationBody(this.SharedState.CurrentCallable!, generatedContent, lambdaParams, callableInfo, isReturnStatement, out call, out callable);
```

In what situations can lifting fail?

---

For lifting in general, a block of code will fail to be lifted (or in other words is not liftable) if it contains a set statement on a mutable defined outside of the block of code, or if it either contains a return statement when not expecting a return or does not contain a return statement when expecting a return, as determined by the parameter `isReturnsAllowed`. It will also fail if multiple return statements are found that disagree on the return type. Most of this is beyond the scope of the lambda lifter as it only lifts a single expression.

---

In that sense, maybe the compiler should throw an error if `success` is false, as I don't think it should be possible for a Q# writer to write a lambda that can't be lifted. Failure here should be an indicator of the compiler failing on some level.

---

Yeah, I think `success` being false is an internal compiler error, so throwing seems better.

---

I've changed this to throw an error on failure.